x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/threat_marker/scan.ts

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import type { Client } from '@elastic/elasticsearch';
+import async from 'async';
+
+export interface Threat {
+  'threat.indicator.type': string;
+  'threat.indicator.url.full': string;
+  'threat.indicator.file.hash.md5': string;
+  'threat.indicator.file.hash.sha1': string;
+}
+
+import type {
+  OpenPointInTimeResponse,
+  SortResults,
+  QueryDslQueryContainer,
+  SearchHit,
+} from '@elastic/elasticsearch/lib/api/types';
+
+/**
+ * Returns filter clause with 'match' conditions for relevant fields only
+ * @param threat
+ * @returns
+ */
+export const shouldClauseForThreat = (threat: Threat) => {
+  switch (threat['threat.indicator.type']) {
+    case 'url': {
+      return [
+        {
+          match: {
+            'url.full': threat['threat.indicator.url.full'],
+          },
+        },
+      ];
+    }
+
+    case 'file': {
+      return [
+        {
+          match: {
+            'file.hash.md5': threat['threat.indicator.file.hash.md5'],
+            'file.hash.sha1': threat['threat.indicator.file.hash.sha1'],
+          },
+        },
+      ];
+    }
+  }
+};
+
+export const THREAT_DETECTION_INDICATOR_FIELD = 'threat.detection.indicator' as const;
+
+export const THREAT_DETECTION_TIMESTAMP_FIELD = 'threat.detection.timestamp' as const;
+
+const updateMapping = async (client: Client, eventsIndex: string[]) => {
+  await client.indices.putMapping({
+    index: eventsIndex,
+    properties: {
+      [THREAT_DETECTION_INDICATOR_FIELD]: {
+        type: 'keyword',
+      },
+      [THREAT_DETECTION_TIMESTAMP_FIELD]: {
+        type: 'date',
+      },
+    },
+  });
+};
+
+export const mark = async (
```

Comment:
I wonder what benefits/drawbacks there are in turning the logic around. Instead of marking source events with threats, have a counter on the indicators documents and increment it when there is a match. Is it feasible? The reason I ask is that if we do it like that we won't need changes in the IoC querying logic, though I'm not sure how well it will play long term, with grouping and such